packages/data-broker/src/services/dhis/DhisService.js

```diff
@@ -226,27 +197,81 @@ export class DhisService extends Service {
     return buildAnalyticsFromEvents(translatedEvents, dataElements);
   };
 
-  pullAnalyticsFromEventsForApi = async (api, dataSources, options) => {
-    const { programCodes = [], period, startDate, endDate } = options;
+  translateDhisEventAnalytics = async (dhisAnalytics, dataSources) => {
     const dataElementCodes = dataSources.map(({ code }) => code);
-    const dhisElementCodes = dataSources.map(({ dataElementCode }) => dataElementCode);
 
-    const query = {
-      programCodes,
-      dataElementCodes: dhisElementCodes,
+    // Map to translation
+    const translatedEventAnalytics = await this.translator.translateInboundEventAnalytics(
+      dhisAnalytics,
+      dataSources,
+    );
+
+    // Map to analytics
+    const analytics = buildAnalyticsFromDhisEventAnalytics(
+      translatedEventAnalytics,
+      dataElementCodes,
+    );
+
+    return analytics;
+  };
+
+  getDataSourcesInProgram = async (dataSources, programCode) => {
+    const allowedDataSources = await this.models.dataSource.getDataElementsInGroup(programCode);
+    const allowedDataSourceCodes = allowedDataSources.map(({ code }) => code);
+    return dataSources.filter(({ code }) => allowedDataSourceCodes.includes(code)); // What check, dataElementCode, code, id?
+  };
+
+  mergeAnalytics = analytics => ({
+    results: analytics
+      .map(({ results }) => results)
+      .flat()
+      .sort(getSortByKey('period')),
+    metadata: {
+      dataElementCodeToName: analytics.reduce(
+        (dataElementCodeToName, { metadata }) => ({
+          ...dataElementCodeToName,
+          ...(metadata.dataElementCodeToName || {}),
+        }),
+        {},
+      ),
+    },
+  })
+
+  pullAnalyticsFromEventsForApi = async (api, dataSources, options) => {
+    const { programCodes = [], period, startDate, endDate, organisationUnitCodes } = options;
+
+    const baseQuery = {
+      // programCodes,
+      organisationUnitCodes,
       dataElementIdScheme: 'code',
       period,
       startDate,
       endDate,
     };
-    const eventAnalytics = await this.fetchEventAnalyticsForPrograms(api, programCodes, query);
 
-    const translatedEventAnalytics = await this.translator.translateInboundEventAnalytics(
-      eventAnalytics,
-      dataSources,
-    );
+    const allAnalytics = [];
 
-    return buildAnalyticsFromDhisEventAnalytics(translatedEventAnalytics, dataElementCodes);
+    const fetchAnalyticsForProgram = async programCode => {
+      const dataSourcesToFetch = await this.getDataSourcesInProgram(dataSources, programCode);
+
+      if (dataSourcesToFetch.length === 0) return;
```

Comment:
Is stopping the fetching process if there are no `dataSources` to fetch going to have any unintended side affects?